### PR TITLE
Add option to store the hash of the upsert primary key

### PR DESCRIPTION
**Pull request by yupeng9**
Add an upsert config `hashFunction` to store the hash value instead of the primary key in the upsert metadata manager for tracking. This is an optimization for memory usage when the given primary key is complex or contains large values (e.g. a composite primary key consisting of multple UUIDs).  Currently the available hash functions are Murmur3 and MD5, both return a 128-bit hash value. 

https://softwareengineering.stackexchange.com/questions/49550/which-hashing-algorithm-is-best-for-uniqueness-and-speed is a good post to read about the hash functions.

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java [hunk 1]
```diff
@@ -35,6 +36,10 @@ public PrimaryKey(Object[] values) {
     return _values;
   }
 
+  public byte[] asBytes() {
+    return SerializationUtils.serialize(_values);
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java:40]: I think this util uses the java ser/de to serialize the values. Can it serialize the `ByteArray` values?
Java ser/de is usually not efficient, so we might want to build our own custom serializer.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 2]
```diff
@@ -40,9 +40,16 @@
     APPEND, INCREMENT, OVERWRITE, UNION
   }
 
+  public enum HashFunction {
+    NONE, MD5, MURMUR3
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:44]: Suggest using `null` to represent disabling hashing. We don't expect people to explicitly put `NONE` as the hash function.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java [hunk 3]
```diff
@@ -239,11 +238,12 @@ public void removeSegment(IndexSegment segment) {
...
 
   public static final class RecordInfo {
-    private final PrimaryKey _primaryKey;
+    /** stores the primary key of the record or its hash value */
+    private final Object _primaryKey;
     private final int _docId;
     private final Comparable _comparisonValue;
 
-    public RecordInfo(PrimaryKey primaryKey, int docId, Comparable comparisonValue) {
+    public RecordInfo(Object primaryKey, int docId, Comparable comparisonValue) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java:246]: We should do the hash within the metadata manager instead of on the caller side. The primary key handling logic should all be wrapped within this class

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/HashUtils.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+  public static byte[] hashMurmur3(byte[] bytes) {
+    return Hashing.murmur3_128().hashBytes(bytes).asBytes();
+  }
+
+  public static byte[] hashMD5(byte[] bytes) {
+    return Hashing.md5().hashBytes(bytes).asBytes();
+  }
+
+  public static Object hashPrimaryKey(PrimaryKey primaryKey, UpsertConfig.HashFunction hashFunction) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/HashUtils.java:39]: This method is specific for upsert handling, so I'd suggest move it out of this general util class.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 2 above)
Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:44]: it's easier for reading the serialized value

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java (hunk 1 above)
Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java:40]: any issue with using java ser/de? I thought the pk is of primitive values, so is there a need to build a custom serializer?

Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java:40]: Primary key might contain `ByteArray` which is not `Serializable`.
Can we check the byte size for the java ser/de result? My impression is that it will be quite large. This will be a very frequent operation for upsert (per record), so want to make sure it is efficient

Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java:40]: oh, no, this is from Apache commons lib which can serialize any object. Let me add a test case on ByteArray as an example.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:44]: Why is that easier? You still need to perform `null` check right?
My concern is that:
1. `NONE` is not really a valid `HashFunction`
2. We usually use `null` to represent that a config is not available in most places. It might be a little bit confusing to use `NONE` just for upsert

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java:40]: I checked the internal implementation, and it is leveraging the `ObjectOutputStream` which uses the java ser/de.
Yeah, let's add a test to make sure it works as expected

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java (hunk 3 above)
Comment (yupeng9) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java:246]: I thought about it. but this way makes the intermediate structure `RecordInfo` even lighter

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 2 above)
Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:44]: well, we have `NONE` as upsert mode value?

Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:44]: reading is easier. When I look at the table config, I can view the full info, and dont have to infer if hash function is enabled or not

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java (hunk 1 above)
Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java:40]: you are right. i need to mark ByteArray as serializable. Any other types that we need to consider?

Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java:40]: Other types should be fine

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:44]: That one I also suggested using `null` in the code review... Having an upsert config with `NONE` mode which disables it doesn't seems very clean to me. But I'm okay if you strongly feel this way is better

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java:246]: `RecordInfo` should wrap the necessary information, and it's up to the manager to process the info instead of the caller. It doesn't make `RecordInfo` lighter because both of them are just one reference

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java [hunk 5]
```diff
@@ -374,10 +376,12 @@ public long getLatestIngestionTimestamp() {
       _validDocIds = new ThreadSafeMutableRoaringBitmap();
       String upsertComparisonColumn = config.getUpsertComparisonColumn();
       _upsertComparisonColumn = upsertComparisonColumn != null ? upsertComparisonColumn : _timeColumnName;
+      _hashFunction = config.getHashFunction();
     } else {
       _partitionUpsertMetadataManager = null;
       _validDocIds = null;
       _upsertComparisonColumn = null;
+      _hashFunction=null;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:384]: (nit) reformat

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/HashUtils.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/HashUtils.java:39]: Let's move this method into the `PartitionUpsertMetadataManager` and keep the primary key handling logic within it

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java [hunk 6]
```diff
@@ -19,6 +19,7 @@
 package org.apache.pinot.spi.data.readers;
 
 import java.util.Arrays;
+import org.apache.commons.lang.SerializationUtils;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java:22]: Use the utils from `lang3`

File: pinot-spi/src/test/java/org/apache/pinot/spi/data/readers/PrimaryKeyTest.java [hunk 7]
```diff
@@ -18,6 +18,8 @@
  */
 package org.apache.pinot.spi.data.readers;
 
+import org.apache.commons.lang.SerializationUtils;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/test/java/org/apache/pinot/spi/data/readers/PrimaryKeyTest.java:21]: Use `lang3`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java (hunk 3 above)
Comment (yupeng9) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java:246]: hmm, to check we are speaking about the same thing.

do you mean
```
class RecordInfo {
    private final Object _primaryKey;
    private final int _docId;
    private final Comparable _comparisonValue;
   private final HashFunction _function;
```
by carrying additional field?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/HashUtils.java (hunk 4 above)
Comment (yupeng9) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/HashUtils.java:39]: Right, this depends on where to put the logic of hashing the primary key.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java:246]: No, `HashFunction` should be a member variable of `PartitionUpsertMetadataManager` which can be passed in from the constructor

Comment (yupeng9) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java:246]: I see. That also works

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/RealtimeSegmentConfig.java [hunk 8]
```diff
@@ -69,7 +70,7 @@ private RealtimeSegmentConfig(String tableNameWithType, String segmentName, Stri
       RealtimeSegmentZKMetadata realtimeSegmentZKMetadata, boolean offHeap, PinotDataBufferMemoryManager memoryManager,
       RealtimeSegmentStatsHistory statsHistory, String partitionColumn, PartitionFunction partitionFunction,
       int partitionId, boolean aggregateMetrics, boolean nullHandlingEnabled, String consumerDir,
-      UpsertConfig.Mode upsertMode, String upsertComparisonColumn,
+      UpsertConfig.Mode upsertMode, String upsertComparisonColumn, UpsertConfig.HashFunction hashFunction,
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/RealtimeSegmentConfig.java:73]: I believe we can revert the change in this file because hashing is handled in the upsert metadata manager

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/RealtimeSegmentConfig.java (hunk 8 above)
Comment (yupeng9) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/RealtimeSegmentConfig.java:73]: true. But I feel it might be good to carry this info, in case we need it in future.
